egs/wsj/s5/steps/cleanup/append_eval_to_ctm.py

```diff
@@ -0,0 +1,396 @@
+
+        i = len(ctm_appended) - 1
+        while i >= 0 and ctm_appended[i][4] == self.silence_symbol:
+            assert(ctm_appended[i][-1] in kSilenceEdits)
+            if ctm_appended[i][-1] == 'EC':
+                ctm_appended[i] = ctm_appended[i][0:-1] + ('EI',)
+            i -= 1
+
+        while (ctm[ctm_iter][4] == self.silence_symbol):
+            ctm_appended.append(ctm[ctm_iter] +
+                                (self.silence_symbol, 'EI'))
+            ctm_iter += 1
+            assert(ctm_iter < len(ctm))
+
+        # hyp word must be in the CTM
+        assert (ctm[ctm_iter][4] == eval_vec[eval_iter][1])
+
+        # Add silence_symbol as the reference word. This will probably not be
+        # used anyway, so its ok to not use special_symbol
+        ctm_appended.append(ctm[ctm_iter] +
+                (self.silence_symbol, 'I'))
+        ctm_iter += 1
+
+        while ctm_iter < len(ctm) and ctm[ctm_iter][4] == self.silence_symbol:
+            ctm_appended.append(ctm[ctm_iter] +
+                                (self.silence_symbol, 'EI'))
+            ctm_iter += 1
+
+        eval_iter += 1
+
+        return ctm_iter, eval_iter
+
+    # Process a substitution in the aligned text file.
+    # Processing steps are as follows:
+    # 1) All the silences in the CTM just before this that were marked as 'EC'
+    # or 'EI' will be changed to 'ES'
+    # 2) All the silences in the CTM follwing this will be marked 'ES' with the
+    # reference word as silence_symbol
+    # 3) The line corresponding to the substituted hyp word, will be marked 'S'
+    # with the reference word as ref word
+    # 4) The silences following the hyp word will be marked 'ES' with the
+    # reference word as silence_symbol
+    def ProcessSubstitution(self, ctm, eval_vec, ctm_iter, eval_iter,
+            ctm_appended):
+        assert(ctm_iter < len(ctm))
+        assert(eval_iter < len(eval_vec))
+
+        i = len(ctm_appended) - 1
+        while i >= 0 and ctm_appended[i][4] == self.silence_symbol:
+            assert(ctm_appended[i][-1] in kSilenceEdits)
+            if ctm_appended[i][-1] in [ 'EI', 'EC' ]:
+                ctm_appended[i] = ctm_appended[i][0:-1] + ('ES',)
+            i -= 1
+
+        while (ctm[ctm_iter][4] == self.silence_symbol):
+            ctm_appended.append(ctm[ctm_iter] +
+                                (self.silence_symbol, 'ES'))
+            ctm_iter += 1
+            assert(ctm_iter < len(ctm))
+
+        # hyp word must be in the CTM
+        assert(ctm[ctm_iter][4] == eval_vec[eval_iter][1])
+
+        if (ctm[ctm_iter][4] == self.oov_symbol and
+            ( (eval_iter > 0 and eval_vec[eval_iter-1] == 'C')
+                or (eval_iter < len(eval_vec)-1 and eval_vec[eval_iter+1] == 'C')
+                or eval_iter == 0 or eval_iter == len(eval_vec)-1 )
+            ):
+            # Substitution by an OOV is treated specially
```

Comment:
you need to distinguish between 'an OOV', which would mean a specific word that is not in the vocabulary (e.g. 'zydeco'), and the OOV symbol.  It's not clear here which you mean.  Have words that are out of the vocabulary been replaced at some point with the unknown-word-symbol.